drivers: ethernet: sam: Add support to GMAC1

TonyHan11 opened the pull request:
There are 2 network interfaces on SAMA7G54-EK board. KSZ9131 is used for GMAC0 (1Gbps) and KSZ8081 is used for GMAC1 (100Mbps). This PR adds the support for GMAC1 first as KSZ8081 is already supported in the mainline.
Changes in this PR includes:
 - configure MMU and clocks for GMAC0 & GMAC1
 - update the cache management for SAMA7G5
 - add GMAC1 node to the device tree
 - add RGMII mode support
 - add GMAC reference clock support

FYI: KSZ9131 support please refer to https://github.com/zephyrproject-rtos/zephyr/pull/90711

TonyHan11 committed:
soc: microchip: sam: update MMU for sama7g5 GMAC1

Enable strong ordered access to the GMAC1 registers.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: ethernet: sam_gmac: update for support sama7g54

Use proper settings of clock, data bus width for sama7g54.
Update queue number checking and the macro definition in header file.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: ethernet: sam_gmac: add GMAC Reference Clock support

Add property for selecting the source for GMAC Reference Clock to dts
bindings yaml file.
Choose the source for the GMAC Reference Clock by GMAC_UR register.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
soc: microchip: sam: configure the clocks for sama7g5 GMAC1

Configure ETH PLL to 625MHz.
Configure the generic clocks to 125MHz for GMAC1.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
dts: microchip: sam: add nodes for GMAC1 to sama7g5.dtsi

Add gmac1 and gmac1_mdio nodes.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
boards: microchip: sam: add nodes for GMAC1 to sama7g54_ek dts file

Add gmac1 and gmac1_mdio nodes.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: ethernet: sam_gmac: use sys_cache_data_foo() for coherence

Replace cache coherence functions:
 - SCB_InvalidateDCache_by_Addr() to sys_cache_data_invd_range()
 - SCB_CleanDCache_by_Addr() to sys_cache_data_flush_range()

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: mdio: sam: update clocks for supporting sama7g54

Do not config the clocks as they are already done in soc.c.

Signed-off-by: Tony Han <tony.han@microchip.com>

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -66,41 +67,50 @@ static inline void dcache_is_enabled(void)
+#ifdef CONFIG_DCACHE
+#undef dcache_invalidate
+#undef dcache_clean
+#define dcache_invalidate sys_cache_data_invd_range
```
why not use sys_cache_data_invd_range for all? as the implementation in zephyr for the cortex_m uses SCB_InvalidateDCache_by_Addr internally

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -19,7 +19,7 @@
  */
 
-#if defined(CONFIG_SOC_FAMILY_ATMEL_SAM)
+#if defined(CONFIG_SOC_FAMILY_ATMEL_SAM) || defined(CONFIG_SOC_SAMA7G54)
```
```suggestion
#if DT_HAS_COMPAT_STATUS_OKAY(atmel_sam_gmac)
```

maass-hamburg commented on dts/arm/microchip/sam/sama7g5.dtsi:
```diff
@@ -501,5 +501,27 @@
+			interrupts = <GIC_SPI 52 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
+				     <GIC_SPI 121 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+			interrupt-names = "gmac", "q1";
+			clocks = <&pmc PMC_TYPE_GCK 52>;
```
this prop also seems to not been used

maass-hamburg commented on boards/microchip/sam/sama7g54_ek/pre_dt_board.cmake:
```diff
@@ -0,0 +1,9 @@
+
+# Suppress "unique_unit_address_if_enabled" to handle the following overlaps:
+# - /soc/ethernet@e2800000 & /soc/mdio@e2800000
+# - /soc/ethernet@e2804000 & /soc/mdio@e2804000
```
if they share registers, its common to have a common parent node, like the stm32 ethernet has

TonyHan11 commented:
@maass-hamburg Thank you for your comments. Rebased and updated with the following changes:
 - use sys_cache_data_xxx() functions for the cache coherence
 - replace `#if defined(CONFIG_SOC_FAMILY_ATMEL_SAM) || defined(CONFIG_SOC_SAMA7G54)` with `#if DT_HAS_COMPAT_STATUS_OKAY(atmel_sam_gmac)`
 - Use a common parent node for ethernet mac and mdio

TonyHan11 committed:
board: microchip: sam: sama7g54_ek enable NET_L2_ETHERNET

Enable NET_L2_ETHERNET in Kconfig.defconfig.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts:
```diff
@@ -119,6 +119,29 @@
 	};
 };
 
+&gmac1 {
```
Yes. If there are in a separated PR, the number of commits will not be too many. 
By deleting the commits for GMAC0, the commits are reduced.

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -31,6 +31,7 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
+#include <zephyr/cache.h>
```
Rename the commit brief changing:

```diff
-sys_cache_data_xxxx()
+sys_cache_data_foo()
```

This is to avoid malicious language problems.

TonyHan11 replied:
Updated the commit message with `sys_cache_data_xxxx()` replaced by `sys_cache_data_foo()`, thanks a lot.

nandojve reviewed (changes requested):
LGTM, just an small nit

TonyHan11 commented:
The commits have conflicts due to the sdmmc is merged.
Rebased with conflicts solved (the dtsi node in `dts/arm/microchip/sam/sama7g5.dtsi` and clock configuration in `soc/microchip/sam/sama7g5/soc.c`).

TonyHan11 commented:
Hi @nandojve ,
I've rebased the branch with conflicts solved. Could you kindly review and approve it again? Thanks a lot.

nandojve commented:
Hi @TonyHan11 ,

Needs rebase!

TonyHan11 commented:
> Hi @TonyHan11 ,
> 
> Needs rebase!

Rebased, thanks.
